### PR TITLE
[FEATURE] Ajouter un bouton "Détacher" dans la liste des filles d'une organisation mère (PIX-17882)

### PR DESCRIPTION
**Pull request by ThomasBazin**
## 🍂 Problème

Dans le cadre de la simplification de la gestion des réseaux d'organisations, on veut pouvoir détacher une organisation fille de son organisation mère.

## 🌰 Proposition

Dans Pix Admin, sur la page d'une organisation mère, dans l'onglet **Organisations filles**, ajouter une colonne _Actions_ dans le tableau, et un bouton _Détacher_ sur chaque ligne d'organisation fille. Une confirmation est demandée avant d'effectuer l'opération de "détachement".

## 🍁 Remarques

Ce bouton n'est visible que pour le rôle Super Admin

## 🪵 Pour tester

Sur Pix Admin:
- se connecter en tant que Super Admin
- aller sur la page d'une organisation mère
- aller sur l'onglet **Organisations filles**
- tenter de détacher une ou plusieurs filles grâce au nouveau bouton _Détacher_

- tenter de se connecter avec un autre rôle 
- constater que la colonne _Actions_ et le bouton _Détacher_ ne sont pas visibles

### CODE REVIEW

File: admin/tests/unit/adapters/organization-test.js [hunk 1]
```diff
@@ -84,4 +84,22 @@ module('Unit | Adapters | organization', function (hooks) {
...
+      // given
+      const childOrganizationId = '1234';
+
+      // when
+      await adapter.detachChildOrganizationFromParent({ childOrganizationId });
+
+      // then
+      assert.true(
+        adapter.ajax.calledOnceWithExactly(
+          'http://localhost:3000/api/admin/organizations/1234/detach-parent-organization',
```
Comment (Alexandre-Monney) [inline admin/tests/unit/adapters/organization-test.js]: **todo**: remplacer le http://localhost:3000 par la var d'en `ENV.APP.API_HOST`

File: admin/tests/unit/controllers/authenticated/organizations/get/children-test.js [hunk 2]
```diff
@@ -168,4 +168,105 @@ module('Unit | Controller | authenticated/organizations/get/children', function
...
+
+      controller.model = { organization: parentOrganization };
+
+      reloadStub = sinon.stub();
+      controller.model.organization.hasMany = sinon.stub().returns({ reload: reloadStub });
+    });
+
+    test('it should detach child organization from parent and display success notification', async function (assert) {
+      // given
+      const detachChildOrganizationFromParent = sinon.stub().resolves();
```
Comment (Alexandre-Monney) [inline admin/tests/unit/controllers/authenticated/organizations/get/children-test.js]:
```suggestion
      const detachChildOrganizationFromParentStub = sinon.stub().resolves();

```

Comment (Alexandre-Monney) [reply admin/tests/unit/controllers/authenticated/organizations/get/children-test.js]: **nitpick**: On peut faire le stub dans le beforeEach, et ensuite faire le resolves/rejects dans le test associé

File: admin/app/components/organizations/children/list-item.gjs [hunk 3]
```diff
@@ -1,32 +1,57 @@
...
-    <:cell>
-      {{@organization.externalId}}
-    </:cell>
-  </PixTableColumn>
-</template>
+import ActionsOnChildrenInOrganization from './actions-on-children-in-organization';
+
+export default class ListItem extends Component {
+  @service accessControl;
+  <template>
```
Comment (Alexandre-Monney) [inline admin/app/components/organizations/children/list-item.gjs:12]: **nitpick**: un petit espace ici sivouplé

File: admin/tests/integration/components/organizations/children/actions-on-children-in-organization-test.gjs [hunk 4]
```diff
@@ -0,0 +1,124 @@
+import { render as renderScreen, within } from '@1024pix/ember-testing-library';
```
Comment (Alexandre-Monney) [inline admin/tests/integration/components/organizations/children/actions-on-children-in-organization-test.gjs]: **thought**: Pas sur qu'on ai besoin du renommage ici

File: admin/tests/integration/components/organizations/children/list-item-test.gjs [hunk 5]
```diff
@@ -1,26 +1,67 @@
...
+
+      // then
+      assert.dom(screen.getByRole('cell', { name: '1' })).exists();
+      assert.dom(screen.getByRole('cell', { name: 'Collège The Night Watch' })).exists();
+      assert.dom(screen.getByRole('cell', { name: 'UA123456' })).exists();
+      assert.ok(
+        within(
+          screen.getByRole('cell', { name: t('components.organizations.children-list.table-headers.actions') }),
+        ).getByRole('button', { name: t('components.organizations.children-list.actions.detach.button') }),
+      );
```
Comment (Alexandre-Monney) [inline admin/tests/integration/components/organizations/children/list-item-test.gjs:36]: **thought**: Peut être simplifier en n'allant chercher que le bouton détacher, on est déjà dans un composant qui gère les lignes d'un tableau

File: admin/tests/acceptance/authenticated/organizations/get/children-test.js [hunk 6]
```diff
@@ -97,19 +98,55 @@ module('Acceptance | Organizations | Children', function (hooks) {
...
+        });
+        const screen = await visit(`/organizations/${parentOrganization.id}/children`);
+
+        // when
+        const childOrganizationRow = screen.getAllByRole('row')[1];
+        await click(
+          within(childOrganizationRow).getByRole('button', {
+            name: t('components.organizations.children-list.actions.detach.button'),
+          }),
+        );
```
Comment (Alexandre-Monney) [inline admin/tests/acceptance/authenticated/organizations/get/children-test.js]: **though**: Je pense qu'on pourrait juste aller récuperer le ou les bouton(s) détacher. Car la on dépends de l'implémentation du tableau

File: admin/tests/acceptance/authenticated/organizations/get/children-test.js [hunk 7]
```diff
@@ -97,19 +98,55 @@ module('Acceptance | Organizations | Children', function (hooks) {
...
+
+        const modal = await screen.findByRole('dialog');
+        await click(within(modal).getByRole('button', { name: t('common.actions.confirm') }));
+
+        await waitForDialogClose();
+
+        // then
+        assert
+          .dom(screen.getByText(t('pages.organization-children.notifications.success.detach-child-organization')))
```
Comment (Alexandre-Monney) [inline admin/tests/acceptance/authenticated/organizations/get/children-test.js]: **nitpick**: ici soit utiliser comme en dessous un ok/notOk, soit faire deux assert.dom

File: admin/translations/en.json [hunk 8]
```diff
@@ -549,7 +549,15 @@
         }
```
Comment (Alexandre-Monney) [inline admin/translations/en.json:549]: **nitpick**: y'a un petit ; qui s'est glissé dans le nom du commit :D

File: admin/app/services/access-control.js [hunk 9]
```diff
@@ -71,4 +71,8 @@ export default class AccessControlService extends Service {
       this.currentUser.adminMember.isCertif
     );
   }
+
+  get hasAccessToDetachChildOrganizationScope() {
+    return !!this.currentUser.adminMember.isSuperAdmin;
```
Comment (xav-car) [inline admin/app/services/access-control.js:76]: **question** : le isSuperAdmin me fait dire que c'est déjà un boolean. il ne me semble pas util de faire le double `!!`

Comment (xav-car) [reply admin/app/services/access-control.js:76]: le endpoint qui liste les adminMember remonte tout de mémoire. il faudrait ajouter dans les tests le isSuperAdmin false pour les autres cas.

File: admin/app/controllers/authenticated/organizations/get/children.js [hunk 10]
```diff
@@ -60,4 +60,33 @@ export default class AuthenticatedOrganizationsGetChildrenController extends Con
...
+    try {
+      await organizationAdapter.detachChildOrganizationFromParent({ childOrganizationId });
+
+      this.pixToast.sendSuccessNotification({
+        message: this.intl.t('pages.organization-children.notifications.success.detach-child-organization'),
+      });
+
+      await this.model.organization.hasMany('children').reload();
+    } catch (responseError) {
+      const error = get(responseError, 'errors[0]');
```
Comment (La-toile-cosmique) [inline admin/app/controllers/authenticated/organizations/get/children.js]: On vient de virer lodash de orga. 🤷🏼

File: admin/app/components/organizations/children/actions-on-children-in-organization.gjs [hunk 11]
```diff
@@ -0,0 +1,48 @@
+import PixButton from '@1024pix/pix-ui/components/pix-button';
+import PixModal from '@1024pix/pix-ui/components/pix-modal';
+import { action } from '@ember/object';
+import { service } from '@ember/service';
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { t } from 'ember-intl';
+
+export default class ActionsOnChildrenInOrganization extends Component {
+  @service store;
```
Comment (La-toile-cosmique) [inline admin/app/components/organizations/children/actions-on-children-in-organization.gjs:10]: ca a pas l'air utilisé

File: admin/app/components/organizations/children/actions-on-children-in-organization.gjs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+  @action
+  toggleDisplayConfirmModal() {
+    this.displayConfirmModal = !this.displayConfirmModal;
+  }
+
+  @action
+  async handleDetachChildOrganization() {
+    await this.args.onDetachChildOrganization({ childOrganizationId: this.args.organization.id });
+    this.displayConfirmModal = !this.displayConfirmModal;
```
Comment (La-toile-cosmique) [inline admin/app/components/organizations/children/actions-on-children-in-organization.gjs]: autant la fermer avant si c'est pas conditionné par le succès

File: admin/app/controllers/authenticated/organizations/get/children.js [hunk 13]
```diff
@@ -60,4 +60,33 @@ export default class AuthenticatedOrganizationsGetChildrenController extends Con
...
+
+      let errorMessage = this.intl.t('common.notifications.generic-error');
+
+      if (error?.code === 'UNABLE_TO_DETACH_PARENT_ORGANIZATION_FROM_CHILD_ORGANIZATION') {
+        errorMessage = this.intl.t(
+          'pages.organization-children.notifications.error.unable-to-detach-child-organization',
+        );
+      }
```
Comment (xav-car) [inline admin/app/controllers/authenticated/organizations/get/children.js]: **suggestion**

```
const isErrorCodeHandle = error?.code === 'UNABLE_TO_DETACH_PARENT_ORGANIZATION_FROM_CHILD_ORGANIZATION';

const errorMessageKey = isErrorCodeHandle ? 'pages.organization-children.notifications.error.unable-to-detach-child-organization' : 'common.notifications.generic-error'
```
pixToast => `message: this.intl.t(errorMessageKey)`

File: admin/app/components/organizations/children/actions-on-children-in-organization.gjs (hunk 12 above)
Comment (La-toile-cosmique) [reply admin/app/components/organizations/children/actions-on-children-in-organization.gjs]: d'ailleurs ce serait plus juste d'écrire ```this.displayConfirmModal = false```
Tu veux jamais qu'il ai une autre valeur a ce moment la.

**Review by La-toile-cosmique — APPROVED** [review]

File: admin/tests/unit/controllers/authenticated/organizations/get/children-test.js [hunk 14]
```diff
@@ -168,4 +168,106 @@ module('Unit | Controller | authenticated/organizations/get/children', function
...
+        assert.true(
+          notifications.sendErrorNotification.calledWithExactly({
+            message: this.intl.t('pages.organization-children.notifications.error.unable-to-detach-child-organization'),
+          }),
+        );
+
+        assert.true(reloadStub.notCalled);
+      });
+
+      test('shoud display generic error notification for any other error', async function (assert) {
```
Comment (xav-car) [inline admin/tests/unit/controllers/authenticated/organizations/get/children-test.js]: **suggestion** 

il manque le test ou il n'y a pas de code dans l'erreur, ou qu'il n'y ai pas d'erreur dans le retour ? puisqu'on fait un `error?.code` plus haut

File: admin/app/controllers/authenticated/organizations/get/children.js (hunk 10 above)
Comment (La-toile-cosmique) [reply admin/app/controllers/authenticated/organizations/get/children.js]: Pour etre plus clair, je préconise pas de ne pas l'utiliser, je vous donne juste cette info :-)

File: admin/app/templates/authenticated/organizations/get/children.hbs [hunk 15]
```diff
@@ -11,7 +11,10 @@
     <h2 class="page-section__title">{{t "pages.organization-children.title"}}</h2>
   </header>
   {{#if @model.children}}
-    <Organizations::Children::List @organizations={{@model.children}} />
+    <Organizations::Children::List
+      @organizations={{@model.children}}
+      @onDetachChildOrganization={{this.detachChildOrganizationFromParent}}
+    />
```
Comment (xav-car) [inline admin/app/templates/authenticated/organizations/get/children.hbs]: **suggestion**

si je ne dis pas de betise, le `hbs `peut être transformé en `gjs `dans les templates ( et rapatrié le contenu du controller içi ).

File: admin/tests/acceptance/authenticated/organizations/get/children-test.js [hunk 16]
```diff
@@ -118,6 +154,17 @@ module('Acceptance | Organizations | Children', function (hooks) {
...
+      test('it does not display "Detach" child organization button', async function (assert) {
+        // when
+        const screen = await visit(`/organizations/${parentOrganizationId}/children`);
+
+        // then
+        assert.ok(screen.getByRole('cell', { name: 'Child Orga name' }));
+        assert.notOk(
+          screen.queryByRole('button', { name: t('components.organizations.children-list.actions.detach.button') }),
+        );
+      });
```
Comment (xav-car) [inline admin/tests/acceptance/authenticated/organizations/get/children-test.js]: snif mirage . mais on a pas encore de e2e sur PixAdmin

**Review by xav-car — APPROVED** [review]
tech ok pour ma part

File: admin/app/templates/authenticated/organizations/get/children.hbs (hunk 15 above)
Comment (ThomasBazin) [reply admin/app/templates/authenticated/organizations/get/children.hbs]: tu veux dire rapatrier toute la méthode `detachChildOrganizationFromParent` dans le template (transformé en gjs) ? Qu'est-ce que ça apporterait de ne pas la laisser dans le controller ?

File: admin/app/components/organizations/children/actions-on-children-in-organization.gjs (hunk 12 above)
Comment (ThomasBazin) [reply admin/app/components/organizations/children/actions-on-children-in-organization.gjs]: tout à fait, on ferme la modale dans tous les cas

File: admin/app/templates/authenticated/organizations/get/children.hbs (hunk 15 above)
Comment (xav-car) [reply admin/app/templates/authenticated/organizations/get/children.hbs]: en fait les controller dans ember ne sont util sur pour la pagination. ( avant le gjs on avait pas trop le choix le js du hbs pour le template c'était le controller. 

Comme on tend à s'en séparer. tout ce qui n'est pas de la gestion de params. peut être dans le gjs du template.

Comment (xav-car) [reply admin/app/templates/authenticated/organizations/get/children.hbs]: > en fait les controller dans ember ne sont util que pour l'usage des queryparams ( filter/pagination). ( avant le gjs on avait pas trop le choix le js du hbs pour le template c'était le controller.
> 
> Comme on tend à s'en séparer. tout ce qui n'est pas de la gestion de params. peut être dans le gjs du template.

File: admin/app/services/access-control.js (hunk 9 above)
Comment (ThomasBazin) [reply admin/app/services/access-control.js:76]: oui effectivement, mais il faudrait dans ce cas revoir tout cet access control ainsi que les tests, ce qui dépasse je pense le scope de cette PR. On se le note !

File: admin/app/templates/authenticated/organizations/get/children.hbs (hunk 15 above)
Comment (ThomasBazin) [reply admin/app/templates/authenticated/organizations/get/children.hbs]: on transforme donc le template hbs en gjs, on sort la logique de detach du controller, pour la mettre directement dans le composant `ActionsOnChildrenInOrganization`

File: admin/app/components/organizations/children/actions-on-children-in-organization.gjs (hunk 11 above)
Comment (ThomasBazin) [reply admin/app/components/organizations/children/actions-on-children-in-organization.gjs:10]: de retour pour appeler l'adapter

File: admin/app/components/organizations/children/actions-on-children-in-organization.gjs [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+      @title={{t "components.organizations.children-list.actions.detach.confirm-modal-title"}}
+      @onCloseButtonClick={{this.toggleDisplayConfirmModal}}
+      @showModal={{this.displayConfirmModal}}
+    >
+      <:content>
+        <p>
+          {{t
+            "components.organizations.children-list.actions.detach.confirm-modal-message"
+            organizationId=@childOrganization.id
+            htmlSafe=true
```
Comment (Alexandre-Monney) [inline admin/app/components/organizations/children/actions-on-children-in-organization.gjs]: **thought** : On peut l'enlever, pas de html dans la clé de trad

File: admin/tests/integration/components/organizations/children/actions-on-children-in-organization-test.gjs [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+          await waitForDialogClose();
+
+          // then
+          assert.notOk(
+            screen.queryByRole('dialog', {
+              name: t('components.organizations.children-list.actions.detach.confirm-modal-title'),
+            }),
+          );
+
+          assert.ok(detachChildOrganizationFromParentStub.calledOnce);
```
Comment (Alexandre-Monney) [inline admin/tests/integration/components/organizations/children/actions-on-children-in-organization-test.gjs]: **though**: Pas sur qu'on ai besoin de remettre ces assertions, qui même en cas d'erreur restent inchangées avec le cas passant

File: admin/tests/acceptance/authenticated/organizations/get/children-test.js (hunk 16 above)
Comment (Alexandre-Monney) [inline admin/tests/acceptance/authenticated/organizations/get/children-test.js]: **though**: Pas sur que ce soit pertinent de garder ce test la en acceptance, on fait aucune action, c'est juste de l'affichage (ou non dans ce cas)

**Review by Alexandre-Monney — APPROVED** [review]

**Comment by Alexandre-Monney** [comment]
Func ok 👍
